include/dlaf/communication/kernels/all_reduce.h

```diff
@@ -0,0 +1,109 @@
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+/// @file
+
+#include <mpi.h>
+
+#include "dlaf/common/bag.h"
+#include "dlaf/common/callable_object.h"
+#include "dlaf/common/pipeline.h"
+#include "dlaf/communication/communicator.h"
+#include "dlaf/communication/executor.h"
+#include "dlaf/communication/message.h"
+#include "dlaf/communication/rdma.h"
+#include "dlaf/matrix/tile.h"
+
+namespace dlaf {
+namespace comm {
+
+namespace internal {
+
+template <class T>
+auto allReduce(common::PromiseGuard<comm::Communicator> pcomm, MPI_Op reduce_op,
+               common::internal::Bag<const T> bag_in, common::internal::Bag<T> bag_out,
+               matrix::Tile<const T, Device::CPU> const&, MPI_Request* req) {
+  auto& communicator = pcomm.ref();
+  auto message_in = comm::make_message(hpx::get<1>(bag_in));
+  auto message_out = comm::make_message(hpx::get<1>(bag_out));
+
+  DLAF_MPI_CALL(MPI_Iallreduce(message_in.data(), message_out.data(), message_in.count(),
+                               message_in.mpi_type(), reduce_op, communicator, req));
+
+  return std::move(bag_out);
+}
+
+DLAF_MAKE_CALLABLE_OBJECT(allReduce);
+
+template <class T>
+auto allReduceInPlace(common::PromiseGuard<comm::Communicator> pcomm, MPI_Op reduce_op,
+                      common::internal::Bag<T> bag, MPI_Request* req) {
+  auto& communicator = pcomm.ref();
+  auto message = comm::make_message(hpx::get<1>(bag));
+
+  DLAF_MPI_CALL(MPI_Iallreduce(MPI_IN_PLACE, message.data(), message.count(), message.mpi_type(),
+                               reduce_op, communicator, req));
+
+  return std::move(bag);
+}
+
+DLAF_MAKE_CALLABLE_OBJECT(allReduceInPlace);
+}
+
+template <class T>
+auto scheduleAllReduce(const comm::Executor& ex,
+                       hpx::future<common::PromiseGuard<comm::Communicator>> pcomm, MPI_Op reduce_op,
+                       hpx::shared_future<matrix::Tile<const T, Device::CPU>> tile_in,
+                       hpx::future<matrix::Tile<T, Device::CPU>> tile_out) {
+  hpx::future<common::internal::Bag<const T>> bag_in =
+      hpx::dataflow(hpx::util::unwrapping(common::internal::makeItContiguous_o), tile_in);
+
+  hpx::future<common::internal::Bag<T>> bag_out;
+  {
+    auto wrapped = getUnwrapRetValAndArgs(
+        hpx::dataflow(matrix::unwrapExtendTiles(common::internal::makeItContiguous_o),
+                      std::move(tile_out)));
+    bag_out = std::move(hpx::get<0>(wrapped));
+    tile_out = std::move(hpx::get<0>(std::move(hpx::get<1>(wrapped))));
+  }
+
+  bag_out = hpx::dataflow(ex, hpx::util::unwrapping(internal::allReduce_o), std::move(pcomm), reduce_op,
+                          std::move(bag_in), std::move(bag_out), tile_in);
+
+  tile_out = hpx::dataflow(hpx::util::unwrapping(common::internal::copyBack_o), std::move(tile_out),
+                           std::move(bag_out));
+
+  return tile_out;
+}
```

Comment:
I had a go at trying to formulate this using `let_value` from https://wg21.link/p1897 (but keeping the bag and dataflow etc.). It's definitely at least as verbose as the above, but is in my mind more explicit about keeping various things alive long enough. That might just be because I'm more familiar with the functionality though, so please let me know what you think. Keep in mind that I haven't compiled this, so it may contain bugs:

```
template <class T>
auto scheduleAllReduceSenders(const comm::Executor& ex,
                              hpx::future<common::PromiseGuard<comm::Communicator>> pcomm,
                              MPI_Op reduce_op,
                              hpx::shared_future<matrix::Tile<const T, Device::CPU>> tile_in,
                              hpx::future<matrix::Tile<T, Device::CPU>> tile_out) {
  namespace ex = hpx::execution::experimental;

  hpx::future<common::internal::Bag<const T>> bag_in =
      hpx::dataflow(hpx::util::unwrapping(common::internal::makeItContiguous_o), tile_in);

  // This async scope keeps tile_out alive for the duration of the operation                                                                                                                                                                   
  // returned by the lambda                                                                                                                                                                                                                    
  ex::let_value(std::move(tile_out), [ex, pcomm = std::move(pcomm), reduce_op,
                                      bag_in = std::move(bag_in), tile_in = std::move(tile_in)](
                                         matrix::Tile<T, Device::CPU>& tile_out) {
    auto bag_out = comm::internal::makeItContiguous(tile_out);

    // This async scope keeps tile_in, bag_in, and bag_out alive                                                                                                                                                                               
    // for the duration of the operation returned by the lambda                                                                                                                                                                                
    auto keep_alive = ex::when_all(ex::keep_future(std::move(tile_in)), std::move(bag_in),
                                   ex::just(std::move(bag_out)));
    return ex::let_value(keep_alive, [](hpx::shared_future<matrix::Tile<const T, Device::CPU>>& tile_in,
                                        common::internal::Bag<const T>& bag_in,
                                        common::internal::Bag<T>& bag_out) {
      // Note: no tile_in just for lifetime management, and void future returned                                                                                                                                                               
      hpx::future<void> =
          hpx::dataflow(ex, hpx::util::unwrapping(internal::allReduce_o), std::move(pcomm), reduce_op,
                        std::move(bag_in), std::move(bag_out));
      return hpx::dataflow(hpx::util::unwrapping(common::internal::copyBack_o), std::move(tile_out),
                           std::move(bag_out));
    });
  });
}                                                                                                                                                                                                                                              
```
I've now written out the types fully, but they could be replaced with auto in many places.

Reply:
I should add that I think this could be done without Bag as well, by using `let_value` more extensively.

Reply:
Thanks for pointing out this, it looks more explicit to me too. I quickly read the standard proposal you linked and, if I got it right, `let_value` also returns the same sender we passed in. Is it right?

If yes, this returned future, will be available as soon as the async scope will complete, so it represents a future<Tile> after the operation. What will I get if I pass in a shared_future by copy? I suppose that it will keep alive the Tile and gets passed through the async_scope, so I will get a shared_future back, with the exact same Tile, but that will be ready once the async_scope will finish. It is a trivial question, but just to check if I get it correctly.

Reply:
I'll close this as resolved as this doesn't (and can't) block merging. This was just to start a discussion on how we might do this differently with sender adaptors. Let's continue this when we've made some more progress on using senders/receivers.

Reply:
I seem to have completely forgotten to respond to this.

> `let_value` also returns the same sender we passed in. Is it right?

`let_value` returns a sender that sends the same values as the sender returned from the invocable. Note that I say "a sender that sends the same values" because it is not literally the same sender. The sender returned from the invocable will be internally connected and started, and the values themselves will then be sent to a connected receiver.